### PR TITLE
fix: fix update/delete using subquery only support boolean data type

### LINKED ISSUE #13986: subquery data type in delete statement should be boolean
**Summary**

## Setup
https://github.com/datafuselabs/wizard/blob/main/checksb/sql/updates/bend/setup.sql

## Query

```
-- UPDATE-A2: Mark all 'pending' orders as 'completed' for ETH, where the price is above average.
UPDATE orders
SET status = 'completed'
WHERE asset_type = 'ETH' AND status = 'pending'
AND price > (SELECT AVG(price) FROM orders WHERE asset_type = 'ETH');
```

Error:
```
1001=>subquery data type in delete statement should be boolean
```

This works in snowflake.

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

fix: fix update/delete using subquery only support boolean data type

- Fixes #13986

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14413)

### CODE REVIEW

**Comment by BohuTANG** [comment]
In markdown, this is ok `#issue`, github will link them:
image

File: src/query/sql/src/planner/plans/scalar_expr.rs [hunk 1]
```diff
@@ -663,6 +663,146 @@ pub trait Visitor<'a>: Sized {
...
+        ScalarExpr::LambdaFunction(expr) => visitor.visit_lambda_function(parent, expr),
+        ScalarExpr::FunctionCall(expr) => visitor.visit_function_call(parent, expr),
+        ScalarExpr::CastExpr(expr) => visitor.visit_cast(parent, expr),
+        ScalarExpr::SubqueryExpr(expr) => visitor.visit_subquery(parent, expr),
+        ScalarExpr::UDFServerCall(expr) => visitor.visit_udf_server_call(parent, expr),
+        ScalarExpr::UDFLambdaCall(expr) => visitor.visit_udf_lambda_call(parent, expr),
+    }
+}
+
+pub fn walk_window_with_parent<'a, V: VisitorWithParent<'a>>(
```
Comment (Dousir9) [inline src/query/sql/src/planner/plans/scalar_expr.rs:776]: It's better to move these code to `visit_window_function`.

File: src/query/sql/src/planner/binder/delete.rs [hunk 2]
```diff
@@ -195,21 +216,27 @@ impl Binder {
...
+                subquery: &'a SubqueryExpr,
+            ) -> Result<()> {
+                self.subqueries.push((parent, subquery));
                 Ok(())
             }
         }
 
         let mut find_subquery = FindSubqueryVisitor { subqueries: vec![] };
-        find_subquery.visit(scalar)?;
+        find_subquery.visit(Some(scalar), scalar)?;
```
Comment (Dousir9) [inline src/query/sql/src/planner/binder/delete.rs:234]: `find_subquery.visit(None, scalar)?;`

**Review by Dousir9 — APPROVED** [review]

File: src/query/sql/src/planner/plans/scalar_expr.rs [hunk 3]
```diff
@@ -663,6 +663,156 @@ pub trait Visitor<'a>: Sized {
     }
 }
 
+pub trait VisitorWithParent<'a>: Sized {
```
Comment (Dousir9) [inline src/query/sql/src/planner/plans/scalar_expr.rs:666]: Rename it to `VisitorWithRootScalarExpr` may improve the readability of the code

File: src/query/sql/src/planner/plans/scalar_expr.rs [hunk 4]
```diff
@@ -663,6 +663,156 @@ pub trait Visitor<'a>: Sized {
...
+    fn visit(&mut self, expr: &'a ScalarExpr) -> Result<()> {
+        walk_expr_with_parent(self, Some(expr), expr)
+    }
+
+    fn visit_with_parent(
+        &mut self,
+        parent: Option<&'a ScalarExpr>,
```
Comment (Dousir9) [inline src/query/sql/src/planner/plans/scalar_expr.rs:673]: we can remove `Option`.

**Review by xudong963 — APPROVED** [review]

**Pull request merged**